### ISSUE TITLE
Very slow process of building secondary indexes

### ISSUE THREAD

**Issue by ooliynyk** [Stratio/cassandra-lucene-index#237]
Hello community.

We have cassandra 3.7 cluster with 4 nodes in which we store terabytes of genetic information.
We manually generate sstables and upload them to the cluster through sstableloader.

The problem is that the compaction process of building Lucene indexes is very slow (less than one percent per day). Also note that Lucene index stored in SSD drives, but it seems that they are built in directories where Cassandra store data.

```
./nodetool compactionstats
pending tasks: 34
- a.allele_id: 4
- a.allele_chrom_pos: 1
- a.variants: 29

id                                   compaction type       keyspace table    completed total     unit  progress
ba057490-8e01-11e6-8d93-3d94cce91789 Secondary index build a variants 67693827  160654845 bytes 42.14%  
e44afd10-8e01-11e6-8d93-3d94cce91789 Secondary index build a variants 12835926  160144834 bytes 8.02%   
Active compaction remaining time :   0h00m00s
```

Our schema: 
```
CREATE KEYSPACE a WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '2'}  AND durable_writes = true;
...
CREATE TABLE a.variants (
    chrom text,
    pos bigint,
    id text,
    file_name text,
    user_id int,
    alt text,
    filter text,
    format text,
    info text,
    is_child boolean,
    qual float,
    ref text,
    PRIMARY KEY ((chrom, pos, id), file_name, user_id)
) WITH CLUSTERING ORDER BY (file_name ASC, user_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
CREATE CUSTOM INDEX variant_idx ON a.variants () USING 'com.stratio.cassandra.lucene.Index' WITH OPTIONS = {'schema': '{
                fields : {
                    chrom : {type : "string"},
                    pos : {type : "long"},
                    alt : {type : "string"},
                    id : {type : "string"},
                    ref : {type : "string"},
                    file_name : {type : "string"}
                }
            }', 'directory_path': '/index/a/variants'};
```

iostat shows that no massive load on the drives
```
avg-cpu:  %user   %nice %system %iowait  %steal   %idle
          26.78    3.81    4.84    8.56    0.00   56.01

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
fd0               0.00         0.00         0.00         84          0
sda              13.79       769.07       805.88 2588550386 2712424612
sdb              35.88      2799.44      2799.65 9422380708 9423102700
sdd               1.30       146.92       456.45  494494930 1536337976
sde               4.49       210.27       315.50  707734841 1061922712
sdi               0.90        57.27       248.45  192756060  836250828
sdf               4.40       102.26       347.05  344197847 1168103712
sdh               1.30       149.50       456.09  503198822 1535104112
sdg              17.08      1439.34       289.94 4844548806  975877172
sdj              11.33      1025.98       345.93 3453256582 1164333880
sdk               1.31       153.88       456.06  517932398 1535017620
sdl              13.13      1178.76       286.89 3967499898  965624892
sdm               1.30       149.51       456.07  503230290 1535042084
sdc               1.30       149.44       456.14  502989718 1535282600
dm-0             50.60      4660.86      3767.53 15687579197 12680793876
dm-1              4.40       102.26       347.05  344180093 1168103668
sdn               5.71       338.11      1159.88 1138014844 3903924512
sdo               5.68       337.53      1159.57 1136051660 3902896968
sdp               5.73       337.74      1160.72 1136757432 3906769600
sdq               5.74       338.49      1160.44 1139302468 3905826748
sdr               5.72       338.32      1160.12 1138734028 3904742656
sds               5.74       339.02      1160.69 1141090200 3906670496
sdt               5.61       337.20      1157.97 1134956908 3897499800
sdu               5.64       337.14      1157.97 1134762548 3897526968
```

tablestats
```
./nodetool tablestats a.variants
Keyspace : a
	Read Count: 13340
	Read Latency: 48.24832083958021 ms.
	Write Count: 9523
	Write Latency: 2.1345144387272916 ms.
	Pending Flushes: 0
		Table: variants
		Space used (live): 26500200734706
		Space used (total): 26501170924640
		Space used by snapshots (total): 0
		Off heap memory used (total): 7090345030
		SSTable Compression Ratio: 0.30118903091443405
		Number of keys (estimate): 43935487
		Memtable cell count: 300015
		Memtable data size: 21915715
		Memtable off heap memory used: 154195615
		Memtable switch count: 1
		Local read count: 0
		Local read latency: NaN ms
		Local write count: 4949
		Local write latency: 0.976 ms
		Pending flushes: 0
		Bloom filter false positives: 94354
		Bloom filter false ratio: 0.00286
		Bloom filter space used: 6356055872
		Bloom filter off heap memory used: 6355854464
		Index summary off heap memory used: 311981711
		Compression metadata off heap memory used: 268313240
		Compacted partition minimum bytes: 73
		Compacted partition maximum bytes: 42510
		Compacted partition mean bytes: 476
		Average live cells per slice (last five minutes): 1.4482940573260457
		Maximum live cells per slice (last five minutes): 535
		Average tombstones per slice (last five minutes): 1.0
		Maximum tombstones per slice (last five minutes): 1
		Dropped Mutations: 0

----------------
```

jstack logs [jstack.log.txt](https://github.com/Stratio/cassandra-lucene-index/files/591970/jstack.log.txt)

**Comment by adelapena** [comment]
Hi @ooliynyk,

Genetics is an exciting use case for the Lucene indexes.

Cassandra builds secondary indexes through a "compaction" process that uses a single thread to index the the base table row per row. This probably explain the low resources consumption and the low indexing speed. In order to parallelise the index building, you can use the asynchronous indexing mode by specifying [the number of indexing threads and the size of the indexing queues](https://github.com/Stratio/cassandra-lucene-index/blob/branch-3.7/doc/documentation.rst#indexing). 
```
CREATE CUSTOM INDEX variant_idx ON a.variants () USING 'com.stratio.cassandra.lucene.Index' WITH OPTIONS = {'schema': '{
                fields : {
                    chrom : {type : "string"},
                    pos : {type : "long"},
                    alt : {type : "string"},
                    id : {type : "string"},
                    ref : {type : "string"},
                    file_name : {type : "string"}
                }
            }',
            'indexing_threads': '<two_per_processor>',
            'indexing_queues_size': '<larger_than_the_average_partition_size>',
            'directory_path': '/index/a/variants'};
```
This is specially useful in large boxes as those I guess you are using. The expected indexing speed should be something about a very few thousand rows per second per core. You can use [JMX API](https://github.com/Stratio/cassandra-lucene-index/blob/3.7.0/doc/documentation.rst#jmx-interface) to monitorize the number of indexed documents in each node and do the math to calculate the indexing speed. 

You can also use the option ``ram_buffer_mb``to increase the size of the Lucene writer buffer.

About the placement of the Lucene files, I have tested ``directory_path`` with both [3.7.3](https://github.com/Stratio/cassandra-lucene-index/tree/3.7.3) and [3.7.0](https://github.com/Stratio/cassandra-lucene-index/tree/3.7.0) and it seems to work perfectly. Which specific version of the index plugin are you using with Cassandra 3.7?

I hope that helps,

**Comment by ooliynyk** [comment]
@adelapena, Thanks for your replay

> In order to parallelise the index building, you can use the asynchronous indexing mode by specifying the number of indexing threads and the size of the indexing queues.

Also, whether to increase the number of concurrent compactors in cassandra.yaml?

> About the placement of the Lucene files, I have tested directory_path with both 3.7.3 and 3.7.0 and it seems to work perfectly. 

After building a lucene files are stored in `/index/a/variants`, as it should be. But the process of building indexes is performed in the cassandra data directories, is this normal?
Lucene index plugin in our cluster: `cassandra-lucene-index-plugin-3.7.2-RC1-SNAPSHOT.jar`

> Genetic information makes me think in very large, almost random, strings. Are you indexing this kind of data?

In our case only pos and id have a wide ranges, but they are not unique

**Comment by adelapena** [comment]
Yes, increasing the number concurrent compactor might help, specially if you have SSD disks.

What do you mean with the process of building indexes in the Cassandra data directories? Which kind of files are you seeing in these directories? Can you see a directory named 'lucene'? Compaction is done in C* directories, but the Lucene files should be written to the specified index directory, and the translation from C* rows to Lucene documents should be done in memory.

**Comment by ooliynyk** [comment]
> What do you mean with the process of building indexes in the Cassandra data directories? Which kind of files are you seeing in these directories?

I mean that compaction process `Secondary index build` perform in C* directories.

>  Compaction is done in C* directories
Ok, now understand that the way it should be.

**Comment by ooliynyk** [comment]
Can you please clarify how to calculate `<larger_than_the_average_partition_size>`

**Comment by ooliynyk** [comment]
I just tried to build indexes on 7787545 rows with few configuration.
Seems that there is no difference in performance.

1) building time: ~ 12 min
```
...
schema={
                fields : {
                    chrom : {type : "string"},
                    pos : {type : "long"},
                    alt : {type : "string"},
                    id : {type : "string"},
                    ref : {type : "string"},
                    file_name : {type : "string"}
                }
            }, indexing_threads=32, directory_path=/index/a/variants, ram_buffer_mb=2048, indexing_queues_size=1600, class_name=com.stratio.cassandra.lucene.Index, target=}]]]
INFO  [MigrationStage:1] 2016-11-16 19:16:19,826 Index.java:127 - Getting initialization task of a.variants.variant_idx
INFO  [MigrationStage:1] 2016-11-16 19:16:19,844 Index.java:132 - Index a.variants.variant_idx needs (re)building
INFO  [MigrationStage:1] 2016-11-16 19:16:19,845 Index.java:127 - Getting initialization task of a.variants.variant_idx
INFO  [MigrationStage:1] 2016-11-16 19:16:19,846 Index.java:132 - Index a.variants.variant_idx needs (re)building
INFO  [MemtablePostFlush:6] 2016-11-16 19:16:19,846 Index.java:200 - Flushing Lucene index a.variants.variant_idx
INFO  [SecondaryIndexManagement:3] 2016-11-16 19:16:19,864 FSIndex.java:204 - Truncated variant_idx
INFO  [SecondaryIndexManagement:3] 2016-11-16 19:16:19,864 SecondaryIndexManager.java:358 - Submitting index build of variant_idx for data in BigTableReader(path='/storage/cassandra/data/a/variants-d1819570915f11e69a1f0dda5a927c3f/mb-67-big-Data.db'),BigTableReader(path='/storage/cassandra/data/a/variants-d1819570915f11e69a1f0dda5a927c3f/mb-66-big-Data.db')
INFO  [Service Thread] 2016-11-16 19:21:04,487 GCInspector.java:284 - ConcurrentMarkSweep GC in 603ms.  CMS Old Gen: 2522777960 -> 907749024; Par Eden Space: 4141176 -> 500839840; Par Survivor Space: 24578496 -> 16009920
INFO  [Service Thread] 2016-11-16 19:27:19,050 GCInspector.java:284 - ParNew GC in 218ms.  CMS Old Gen: 2397162680 -> 2486253160; Par Eden Space: 671088640 -> 0;
INFO  [SecondaryIndexManagement:3] 2016-11-16 19:28:04,832 Index.java:200 - Flushing Lucene index a.variants.variant_idx
INFO  [SecondaryIndexManagement:3] 2016-11-16 19:28:25,990 SecondaryIndexManager.java:378 - Index build of variant_idx complete
```

2) building time: ~ 12 min
```
...
schema={
                fields : {
                    chrom : {type : "string"},
                    pos : {type : "long"},
                    alt : {type : "string"},
                    id : {type : "string"},
                    ref : {type : "string"},
                    file_name : {type : "string"}
                }
            }, class_name=com.stratio.cassandra.lucene.Index, directory_path=/index/a/variants, target=}]]]
INFO  [MigrationStage:1] 2016-11-16 19:29:51,227 Index.java:127 - Getting initialization task of a.variants.variant_idx
INFO  [MigrationStage:1] 2016-11-16 19:29:51,228 Index.java:132 - Index a.variants.variant_idx needs (re)building
INFO  [MigrationStage:1] 2016-11-16 19:29:51,228 Index.java:127 - Getting initialization task of a.variants.variant_idx
INFO  [MigrationStage:1] 2016-11-16 19:29:51,229 Index.java:132 - Index a.variants.variant_idx needs (re)building
INFO  [MemtablePostFlush:8] 2016-11-16 19:29:51,229 Index.java:200 - Flushing Lucene index a.variants.variant_idx
INFO  [SecondaryIndexManagement:4] 2016-11-16 19:29:51,245 FSIndex.java:204 - Truncated variant_idx
INFO  [SecondaryIndexManagement:4] 2016-11-16 19:29:51,246 SecondaryIndexManager.java:358 - Submitting index build of variant_idx for data in BigTableReader(path='/storage/cassandra/data/a/variants-d1819570915f11e69a1f0dda5a927c3f/mb-67-big-Data.db'),BigTableReader(path='/storage/cassandra/data/a/variants-d1819570915f11e69a1f0dda5a927c3f/mb-66-big-Data.db')
INFO  [IndexSummaryManager:1] 2016-11-16 19:36:09,248 IndexSummaryRedistribution.java:75 - Redistributing index summaries
INFO  [SecondaryIndexManagement:4] 2016-11-16 19:42:02,437 Index.java:200 - Flushing Lucene index a.variants.variant_idx
INFO  [SecondaryIndexManagement:4] 2016-11-16 19:42:04,856 SecondaryIndexManager.java:378 - Index build of variant_idx complete
```

3) building time: ~ 12 min
```
...
schema={
	fields : {
                    chrom : {type : "string"},
                    pos : {type : "long"},
                    alt : {type : "string"},
                    id : {type : "string"},
                    ref : {type : "string"},
                    file_name : {type : "string"}
                }
            }, indexing_threads=8, directory_path=/index/a/variants, ram_buffer_mb=512, indexing_queues_size=400, class_name=com.stratio.cassandra.lucene.Index, target=}]]]
INFO  [MigrationStage:1] 2016-11-16 19:56:56,509 Index.java:127 - Getting initialization task of a.variants.variant_idx
INFO  [MigrationStage:1] 2016-11-16 19:56:56,509 Index.java:132 - Index a.variants.variant_idx needs (re)building
INFO  [MigrationStage:1] 2016-11-16 19:56:56,509 Index.java:127 - Getting initialization task of a.variants.variant_idx
INFO  [MigrationStage:1] 2016-11-16 19:56:56,509 Index.java:132 - Index a.variants.variant_idx needs (re)building
INFO  [MemtablePostFlush:11] 2016-11-16 19:56:56,510 Index.java:200 - Flushing Lucene index a.variants.variant_idx
INFO  [SecondaryIndexManagement:5] 2016-11-16 19:56:56,527 FSIndex.java:204 - Truncated variant_idx
INFO  [SecondaryIndexManagement:5] 2016-11-16 19:56:56,527 SecondaryIndexManager.java:358 - Submitting index build of variant_idx for data in BigTableReader(path='/storage/cassandra/data/a/variants-d1819570915f11e69a1f0dda5a927c3f/mb-67-big-Data.db'),BigTableReader(path='/storage/cassandra/data/a/variants-d1819570915f11e69a1f0dda5a927c3f/mb-66-big-Data.db')
INFO  [SecondaryIndexManagement:5] 2016-11-16 20:09:03,669 Index.java:200 - Flushing Lucene index a.variants.variant_idx
INFO  [SecondaryIndexManagement:5] 2016-11-16 20:09:09,106 SecondaryIndexManager.java:378 - Index build of variant_idx complete
```

**Comment by adelapena** [comment]
Which hardware are you using?

**Comment by ooliynyk** [comment]
Azure **Standard_DS5_v2** VM's configuration:
**Size	CPU**: 16 cores
**Memory**: 56 GiB
**Max disk throughput:**  51,200 / 768 IOPS / MBps

**Comment by ooliynyk** [comment]
@adelapena, maybe we should wait until the data compaction finished and only then create an lucene indexes?

**Comment by adelapena** [comment]
It will probably help. Could you send us some sample data in order to doing some tests? It would be ideal to have these 7787545 to run some experiments and take a look to the values cardinalities, etc.

**Comment by ooliynyk** [comment]
It would be great. I generated sstabels with sample data, you can download it from gdrive https://drive.google.com/open?id=0B08GeVWJWtcCd0Z6ZWZJMGdMd3c

**Comment by adelapena** [comment]
Hi,

We are testing the dataset. With [this machine](https://www.hetzner.de/hosting/produkte_rootserver/ex51ssd) and eight indexing threads, the index time is around 230 seconds. That is around 33000 documents per second, which is not bad at all. However, without asynchronous indexing the throughput decreases to around 13000 documents per second. Increasing the number of threads produces a very slow decrease of throughput.

Asynchronous indexing is useful to parallelise the translation from rows to documents to be added to Lucene's IndexWriter. The documents are delivered to the [IndexWriter](https://lucene.apache.org/core/5_5_1/core/org/apache/lucene/index/IndexWriter.html) at the end of the indexing queues. The IndexWriter has a single instance per index and it uses its own (limited) parallelism and seems to be bottle neck here, together with the Lucene's segments merger threads.

**Comment by ooliynyk** [comment]
Hi @adelapena,

33000 documents per seconds quite acceptable for us. It seems, that in our case the process was carried out much more slowly.
So I deleted lucene index and waited until all compaction process(data compaction, size estimation etc) was finished. After this, I have created lucene index with configuration that you recommend. Now  process of index building running much faster (about 25% per day compared to 1% per day, what it was before)

Thanks a lot, really appreciate your help!

**Comment by ooliynyk** [comment]
Today I encountered another problem - **number of documents in the index cannot exceed 2147483519**

logs/system.log:

```
...
ERROR [pool-5-thread-1] 2016-12-01 13:42:21,969 IndexException.java:60 - Error indexing Document<indexed,tokenized,omitNorms,indexOptions=DOCS<chrom:chrX> docValuesType=SORTED_SET<chrom:[63 68 72 58]> indexed,tokenized,omitNorms,indexOptions=DOCS,numericType=LONG,numericPrecisionStep=16<pos:88665288> docValuesType=SORTED_NUMERIC<pos:88665288> indexed,tokenized,omitNorms,indexOptions=DOCS<alt:C> docValuesType=SORTED_SET<alt:[43]> indexed,tokenized,omitNorms,indexOptions=DOCS<id:rs184514484> docValuesType=SORTED_SET<id:[72 73 31 38 34 35 31 34 34 38 34]> indexed,tokenized,omitNorms,indexOptions=DOCS<ref:T> docValuesType=SORTED_SET<ref:[54]> indexed,tokenized,omitNorms,indexOptions=DOCS<file_name:HG00103> docValuesType=SORTED_SET<file_name:[48 47 30 30 31 30 33]> indexed,tokenized,omitNorms,indexOptions=DOCS,numericType=LONG,numericPrecisionStep=16,docValuesType=NUMERIC<_token:-793933810096597621> stored,indexed,omitNorms,indexOptions=DOCS,docValuesType=SORTED<_partition_key:[0 4 63 68 72 58 0 0 8 0 0 0 0 5 48 ec c8 0 0 b 72 73 31 38 34 35 31 34 34 38 34 0]> stored,indexed,omitNorms,indexOptions=DOCS,docValuesType=SORTED<_primary_key:[0 10 37 34 66 62 36 31 36 37 66 61 30 35 31 35 38 62 0 0 20 0 4 63 68 72 58 0 0 8 0 0 0 0 5 48 ec c8 0 0 b 72 73 31 38 34 35 31 34 34 38 34 0 0 0 11 0 7 48 47 30 30 31 30 33 0 0 4 0 0 0 0 0 0]>> with term _primary_key:[0 10 37 34 66 62 36 31 36 37 66 61 30 35 31 35 38 62 0 0 20 0 4 63 68 72 58 0 0 8 0 0 0 0 5 48 ec c8 0 0 b 72 73 31 38 34 35 31 34 34 38 34 0 0 0 11 0 7 48 47 30 30 31 30 33 0 0 4 0 0 0 0 0 0] in variant_idx
java.lang.IllegalArgumentException: number of documents in the index cannot exceed 2147483519
        at org.apache.lucene.index.DocumentsWriterPerThread.reserveOneDoc(DocumentsWriterPerThread.java:211) ~[cassandra-lucene-index-plugin-3.7.2-RC1-SNAPSHOT.jar:na]
        at org.apache.lucene.index.DocumentsWriterPerThread.updateDocument(DocumentsWriterPerThread.java:218) ~[cassandra-lucene-index-plugin-3.7.2-RC1-SNAPSHOT.jar:na]
        at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:450) ~[cassandra-lucene-index-plugin-3.7.2-RC1-SNAPSHOT.jar:na]
        at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1477) ~[cassandra-lucene-index-plugin-3.7.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.index.FSIndex.upsert(FSIndex.java:160) ~[cassandra-lucene-index-plugin-3.7.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.IndexService.lambda$upsert$3(IndexService.java:419) [cassandra-lucene-index-plugin-3.7.2-RC1-SNAPSHOT.jar:na]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_101]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_101]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_101]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_101]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_101]
...
```

**Comment by adelapena** [comment]
Hi,

As the message says, Lucene is not able to index more than 2147483519 (MAX_INT - 128). This is [a limit imposed by Lucene](https://issues.apache.org/jira/browse/LUCENE-5843), and there is no much to do here. Although [local index partitioning](https://github.com/Stratio/cassandra-lucene-index/tree/feature/partitioning_on_field_names) is in our roudmap, I encourage you to use more machines with less load to overcome this limitation.

**Comment by ooliynyk** [comment]
Hi, 

> I encourage you to use more machines with less load to overcome this limitation.
The process of building indexes finished less than half (about 45%), and the problem has arisen. If so, we need at least twice as many machines, what is not very acceptable solution.

> Although local index partitioning is in our roudmap
Are these solution compatible with cassandra 3.7?
Also, can you please clarify, whether this solution requires special configuration?

**Comment by adelapena** [comment]
Hi @ooliynyk,

Are your searches going to be partition-directed? If not, have you tested search times with such volume of data per node?

I think local partitioning can be ready in one or two weeks on top of 3.9 branch, and then backported to older Cassandra versions. The idea is to allow to specify a number of partitions at index creation time. Each partition would store only partitions inside a token sub-range. This way, partition-directed queries could be addressed to a specific, smaller index partition, with better performance results. The downside of the approach is that not partition-directed queries should fetch all the index partitions and reconcile the results. Although the origina purpose of this approach was to speed up partition-directed queries in spite of slower not partition-directed queries, it could also solve the limitation in number of documents per Lucene index.

As an alternative, if your searches are partition-directed, you can distribute the data between multiple tables using some domain-specific criteria. For example, if your use case allows it, maybe you can split the data by chromosomes, with one table for chromosomes 1 to 5, 6 to 10, 11 to 15, 16 to 22 , X and Y. Could it fit your use case?

**Comment by ooliynyk** [comment]
> Are your searches going to be partition-directed? If not, have you tested search times with such volume of data per node?

Unfortunately, not quite. Search can be performed on pair of fields **pos** and  **id** fields or on a **chrom** field. Fields **alt** or(and) **ref** in both cases optional.

> If not, have you tested search times with such volume of data per node?

With lucene index is fast enough (usually less than a minute), but that given that our indexes are built only partially.

> As an alternative, if your searches are partition-directed, you can distribute the data between multiple tables using some domain-specific criteria. For example, if your use case allows it, maybe you can split the data by chromosomes, with one table for chromosomes 1 to 5, 6 to 10, 11 to 15, 16 to 22 , X and Y. Could it fit your use case?

As I mentioned above, the search filter may include different fields (not always **chrom**)

**Comment by adelapena** [comment]
Hi @ooliynyk,

I think hopefully we will have a release including index partitioning during this week. Indeed, we already have a working prototype, but we need to run some additional tests and document it.

This way it would be possible to split the index in multiple partitions under the imposed 2147483519 documents limit.

The new feature allows to specify an extensible index partitioner implementation, for example:
```sql
CREATE CUSTOM INDEX wide_idx 
ON test.wide (lucene) 
USING 'com.stratio.cassandra.lucene.Index' 
WITH OPTIONS = {
  ...
  'partitioner': '{type: "token", partitions: 4}',
  ...
};
```
Currently there is only a partitioner implementation that uses the modulus of the hash of the primary key to decide which partition owns each row. This should guarantee a good load balance between partitions. Partition-directed queries are routed to a single index partition, whereas the rest of the queries use a [Lucene MultiReader](https://lucene.apache.org/core/5_3_0/core/org/apache/lucene/index/MultiReader.html) to fetch data from all the partitions. Future implementations will allow to use more custom partitioning criteria.

**Comment by ooliynyk** [comment]
Hi @adelapena,

That sounds good.

In the meantime, we started distributing the data between many segment tables, thus avoiding the imposed documents limit. It allows us to build indexes for all segments separately, which is much more reliable if the node falls.

**Comment by adelapena** [comment]
Nice, that seems a good solution. The new partitioning feature is in [this branch](https://github.com/Stratio/cassandra-lucene-index/tree/feature/partitioning), we hope to merge it very soon.

**Comment by adelapena** [comment]
Hi,

Just to let you know that we have just published [3.7.4](https://github.com/Stratio/cassandra-lucene-index/tree/3.7.4). It includes several [performance improvements](https://github.com/Stratio/cassandra-lucene-index/blob/3.7.4/CHANGELOG.md) that could be useful for your use case. 

Although this new release includes [support for index partitioning](https://github.com/Stratio/cassandra-lucene-index/blob/3.7.4/doc/documentation.rst#partitioners), the only available partitioner implementation is designed to be used with partition-directed searches, so it probably wont be useful for your use case. We will let you know we have an applicable partitioner implementation.

**Comment by ooliynyk** [comment]
Hi, 

Thank you, this is good news.

Our data compaction almost done, soon we will start to build indexes.

**Comment by ooliynyk** [comment]
Hi @adelapena,

> We have just published 3.7.5, 3.8.3, 3.9.5 and 3.0.10.2 with a new column-based partitioner. These new releases are also able of searching in sets of partitions containing more than 2147483519 total rows/documents. This way, the Lucene's limitation in the number of documents per index can be fully overcome by increasing the number of partitions.

This is great! In addition, our decision to splitting the data into segments helped to avoid limitation. Now indexes are built without any problems.

You helped us a lot, thank you very much! I think that the issue can be closed.

**Comment by adelapena** [comment]
These are very good news :) 
Closing.
